aten/src/ATen/NamedTensorUtils.cpp

```diff
@@ -311,81 +312,31 @@ static bool are_distinct(DimnameList batch_dims, DimnameList feature_dims) {
   return true;
 }
 
-// Let batch_dims = everything except for the last two dimensions
-//     feature_dims = the last two dims of the tensor.
-// This function checks that names of batch_dims of one tensor are distinct
-// from the feature dimensions of another tensor.
-//
-// For example,
-// Tensor[N, A, B] @ Tensor[None, N, A] -> Misaligned because N (a batch dim of
-// the first tensor) appears in a non-batch dimension in the second tensor.
-//
-// Tensor[N, A, B1] @ Tensor[N, B2] -> Misaligned! The user may have intended
-// to batch multiply matrices of size [A, B1] with vectors of size [B2]
-// but instead matmul treats it as batch multiplying matrices of size [A, B1]
-// with matrices of size [N, B2]. In this case, we're able to warn the user.
-static void check_batch_and_feature_dims_are_distinct(
-    DimnameList self_names,
-    DimnameList other_names) {
-  auto self_batch_dims = batch_dims(self_names);
-  auto self_feature_dims = feature_dims(self_names);
-  auto other_batch_dims = batch_dims(other_names);
-  auto other_feature_dims = feature_dims(other_names);
-
-  TORCH_CHECK(are_distinct(self_batch_dims, other_feature_dims),
-      "Misaligned dims when batch matrix multiplying Tensor", self_names,
-      " and Tensor", other_names, ": there is overlap between the feature dims ",
-      "of the second tensor (", other_feature_dims,
-      ") and the batch dims of the first tensor (", self_batch_dims,
-      "). Please ensure batch dims are not present in the feature dims.");
-
-  TORCH_CHECK(are_distinct(other_batch_dims, self_feature_dims),
-      "Misaligned dims when batch matrix multiplying Tensor", self_names,
-      " and Tensor", other_names, ": there is overlap between the feature dims ",
-      "of the first tensor (", self_feature_dims,
-      ") and the batch dims of the second tensor (", other_batch_dims,
-      "). Please ensure batch dims are not present in the feature dims.");
-}
-
-// Compute the outnames of torch.matmul(A, B).
+static int64_t num_batch_dims(DimnameList names) {
+  if (names.size() <= 2) {
+    return 0;
+  }
+  return names.size() - 2;
+}
+
 static std::vector<Dimname> compute_matmul_outnames(
     DimnameList self_names,
     DimnameList other_names) {
   TORCH_CHECK(self_names.size() >= 1 && other_names.size() >= 1,
       "both arguments to matmul need to be at least 1D, but they are ",
       self_names.size(), "D and ", other_names.size(), "D");
 
-  check_batch_and_feature_dims_are_distinct(self_names, other_names);
-
-  // The approach is to
-  // (1) compute the outnames of the matrix multiplication on the feature dims.
-  // (2) compute the outnames of the batch dimensions, after broadcasting
-  // (3) concatenate the batch outnames and matrix multiply outnames.
-
-  // Step 1: Compute outnames of matrix multiplication
-  // Let N >= 2.
```

Comment:
I still see `>=2` checks after this PR, should we preserve at least some aspects of the explanation in these comments?

Reply:
Sure. Most of this explanation is about the semantics of matmul (which I am assuming that a future dev who is reading the code understands) but I'll write up a concise explanation

Reply:
Done